Create utils/current.py to list current slow operations

xoto10 opened the pull request:
@tomtor @ppigazzini This script attempts to list the current operations that are slow (seconds_running > 0.3) and with a known query. Perhaps running this occasionally will give us an idea which operations are slow.
It could probably be enhanced to do `uptime` to show the load averages and maybe other diagnostics?

ppigazzini commented:
@xoto10 the first comment must be updated before this can be committed.
@xoto10 @tomtor I already copied by hand the script on the official server. Let me know if you need to know the output.

xoto10 commented:
@ppigazzini @tomtor 
I have updated the comments and also added a call to the os command 'uptime' and made the minimum time a parameter.
I am not sure the conditions on the query are working properly ('seconds running > 0.3s/param value' and 'query != {}'). I suspect this might be due to the different versions of MongoDB in Live and my local fishtest but perhaps I have specified the conditions incorrectly. I suggest you try running the script to see if; a) it runs ok, and b) if the output is of any interest.

ppigazzini commented:
@xoto10 on the official server there is this error message, on the dev server it worked fine.

```
$ python current.py
Current Indexes:
{u'_id_': {u'key': [(u'_id', 1)], u'v': 1},
 u'args.username_1': {u'background': True,
                      u'expireAfterSeconds': 86400,
                      u'key': [(u'args.username', 1)],
                      u'v': 1},
 u'deleted_1': {u'background': True,
                u'expireAfterSeconds': 86400,
                u'key': [(u'deleted', 1)],
                u'v': 1},
 u'deleted_1_finished_1': {u'background': True,
                           u'expireAfterSeconds': 86400,
                           u'key': [(u'deleted', 1), (u'finished', 1)],
                           u'v': 1},
 u'finished_1_last_updated_-1': {u'key': [(u'finished', 1),
                                          (u'last_updated', -1)],
                                 u'v': 1},
 u'finished_1_last_updated_-1_start_time_-1': {u'background': True,
                                               u'expireAfterSeconds': 86400,
                                               u'key': [(u'finished', 1),
                                                        (u'last_updated',
                                                         -1),
                                                        (u'start_time', -1)],
                                               u'v': 1},
 u'last_updated_-1_start_time_-1': {u'background': True,
                                    u'expireAfterSeconds': 86400,
                                    u'key': [(u'last_updated', -1),
                                             (u'start_time', -1)],
                                    u'v': 1},
 u'tasks.pending_1': {u'key': [(u'tasks.pending', 1)], u'v': 1}}

Run 'uptime':

Traceback (most recent call last):
  File "current.py", line 38, in <module>
    printout(subprocess.check_output(["uptime"]))
AttributeError: 'module' object has no attribute 'check_output'
```

xoto10 commented:
Sorry mate, been ill since Saturday, so not reading emails. (Still running
high temperature now.)

Up to you, you could comment out the part that fails, or just wait till I'm
back in action (not in the next day or two).

ppigazzini commented:
@xoto10 yes, now it works

```
Current Indexes:
{u'_id_': {u'key': [(u'_id', 1)], u'v': 1},
 u'args.username_1': {u'background': True,
                      u'expireAfterSeconds': 86400,
                      u'key': [(u'args.username', 1)],
                      u'v': 1},
 u'deleted_1': {u'background': True,
                u'expireAfterSeconds': 86400,
                u'key': [(u'deleted', 1)],
                u'v': 1},
 u'deleted_1_finished_1': {u'background': True,
                           u'expireAfterSeconds': 86400,
                           u'key': [(u'deleted', 1), (u'finished', 1)],
                           u'v': 1},
 u'finished_1_last_updated_-1': {u'key': [(u'finished', 1),
                                          (u'last_updated', -1)],
                                 u'v': 1},
 u'finished_1_last_updated_-1_start_time_-1': {u'background': True,
                                               u'expireAfterSeconds': 86400,
                                               u'key': [(u'finished', 1),
                                                        (u'last_updated',
                                                         -1),
                                                        (u'start_time', -1)],
                                               u'v': 1},
 u'last_updated_-1_start_time_-1': {u'background': True,
                                    u'expireAfterSeconds': 86400,
                                    u'key': [(u'last_updated', -1),
                                             (u'start_time', -1)],
                                    u'v': 1},
 u'tasks.pending_1': {u'key': [(u'tasks.pending', 1)], u'v': 1}}

Run 'uptime':

 00:44:41 up 1 day, 14:04,  1 user,  load average: 0.12, 0.16, 0.21
0

Current operations:
{u'inprog': [{u'active': False,
              u'client': u'127.0.0.1:53758',
              u'connectionId': 2326,
              u'desc': u'conn2326',
              u'lockStats': {u'timeAcquiringMicros': {u'r': 2L, u'w': 0L},
                             u'timeLockedMicros': {u'r': 81L, u'w': 0L}},
              u'ns': u'fishtest_new.users',
              u'numYields': 0,
              u'op': u'query',
              u'opid': 3541598,
              u'query': {u'username': u'ChessDB_CN_WorkerPool'},
              u'threadId': u'0x7fd980996700',
              u'waitingForLock': False},
             {u'active': False,
              u'client': u'127.0.0.1:50550',
              u'connectionId': 2314,
              u'desc': u'conn2314',
              u'lockStats': {u'timeAcquiringMicros': {u'r': 1L, u'w': 0L},
                             u'timeLockedMicros': {u'r': 47L, u'w': 0L}},
              u'ns': u'fishtest_new.users',
              u'numYields': 0,
              u'op': u'query',
              u'opid': 3541286,
              u'query': {u'username': u'ChessDB_CN_WorkerPool'},
              u'threadId': u'0x7fd980592700',
              u'waitingForLock': False},
             {u'active': False,
              u'client': u'127.0.0.1:50616',
              u'connectionId': 2319,
              u'desc': u'conn2319',
              u'lockStats': {u'timeAcquiringMicros': {},
                             u'timeLockedMicros': {}},
              u'ns': u'',
              u'numYields': 0,
              u'op': u'query',
              u'opid': 3629085,
              u'query': {u'getlasterror': 1L},
              u'threadId': u'0x7fd980491700',
              u'waitingForLock': False},
             {u'active': False,
              u'client': u'127.0.0.1:56904',
              u'connectionId': 2333,
              u'desc': u'conn2333',
              u'lockStats': {u'timeAcquiringMicros': {u'r': 45729L,
                                                      u'w': 0L},
                             u'timeLockedMicros': {u'r': 122L, u'w': 0L}},
              u'ns': u'fishtest_new.users',
              u'numYields': 0,
              u'op': u'query',
              u'opid': 3628686,
              u'query': {u'username': u'Bobo1239'},
              u'threadId': u'0x7fd980693700',
              u'waitingForLock': False},
             {u'active': False,
              u'client': u'127.0.0.1:53756',
              u'connectionId': 2325,
              u'desc': u'conn2325',
              u'lockStats': {u'timeAcquiringMicros': {u'r': 1L, u'w': 0L},
                             u'timeLockedMicros': {u'r': 55L, u'w': 0L}},
              u'ns': u'fishtest_new.users',
              u'numYields': 0,
              u'op': u'query',
              u'opid': 3527711,
              u'query': {u'username': u'rkl'},
              u'threadId': u'0x7fd67d8f3700',
              u'waitingForLock': False},
             {u'active': False,
              u'client': u'127.0.0.1:56900',
              u'connectionId': 2331,
              u'desc': u'conn2331',
              u'lockStats': {u'timeAcquiringMicros': {u'r': 32073L,
                                                      u'w': 0L},
                             u'timeLockedMicros': {u'r': 64L, u'w': 0L}},
              u'ns': u'fishtest_new.users',
              u'numYields': 0,
              u'op': u'query',
              u'opid': 3628687,
              u'query': {u'username': u'ChessDB_CN_WorkerPool'},
              u'threadId': u'0x7fd980390700',
              u'waitingForLock': False},
             {u'active': False,
              u'client': u'127.0.0.1:50620',
              u'connectionId': 2321,
              u'desc': u'conn2321',
              u'lockStats': {u'timeAcquiringMicros': {u'r': 34301L,
                                                      u'w': 0L},
                             u'timeLockedMicros': {u'r': 56L, u'w': 0L}},
              u'ns': u'fishtest_new.users',
              u'numYields': 0,
              u'op': u'query',
              u'opid': 3618403,
              u'query': {u'username': u'drabel'},
              u'threadId': u'0x7fd67daf5700',
              u'waitingForLock': False},
             {u'active': False,
              u'client': u'127.0.0.1:50392',
              u'connectionId': 2313,
              u'desc': u'conn2313',
              u'lockStats': {u'timeAcquiringMicros': {u'r': 2L, u'w': 0L},
                             u'timeLockedMicros': {u'r': 49L, u'w': 0L}},
              u'ns': u'fishtest_new.users',
              u'numYields': 0,
              u'op': u'query',
              u'opid': 3606515,
              u'query': {u'username': u'tvijlbrief'},
              u'threadId': u'0x7fd667065700',
              u'waitingForLock': False},
             {u'active': False,
              u'client': u'127.0.0.1:53760',
              u'connectionId': 2327,
              u'desc': u'conn2327',
              u'lockStats': {u'timeAcquiringMicros': {u'r': 23734L,
                                                      u'w': 0L},
                             u'timeLockedMicros': {u'r': 125L, u'w': 0L}},
              u'ns': u'fishtest_new.users',
              u'numYields': 0,
              u'op': u'query',
              u'opid': 3629036,
              u'query': {u'username': u'dew'},
              u'threadId': u'0x7fd67dcf7700',
              u'waitingForLock': False},
             {u'active': False,
              u'client': u'127.0.0.1:50390',
              u'connectionId': 2312,
              u'desc': u'conn2312',
              u'lockStats': {u'timeAcquiringMicros': {u'r': 40634L,
                                                      u'w': 0L},
                             u'timeLockedMicros': {u'r': 113L, u'w': 0L}},
              u'ns': u'fishtest_new.users',
              u'numYields': 0,
              u'op': u'query',
              u'opid': 3503460,
              u'query': {u'username': u'tvijlbrief'},
              u'threadId': u'0x7fd980b98700',
              u'waitingForLock': False},
             {u'active': False,
              u'desc': u'DataFileSync',
              u'lockStats': {u'timeAcquiringMicros': {},
                             u'timeLockedMicros': {}},
              u'ns': u'',
              u'numYields': 0,
              u'op': u'none',
              u'opid': 0,
              u'query': {},
              u'threadId': u'0x7fd98bea3700',
              u'waitingForLock': False},
             {u'active': False,
              u'client': u'0.0.0.0:0',
              u'desc': u'initandlisten',
              u'insert': {},
              u'lockStats': {u'timeAcquiringMicros': {u'r': 0L, u'w': 1L},
                             u'timeLockedMicros': {u'R': 0L, u'W': 122150L}},
              u'ns': u'local.startup_log',
              u'numYields': 0,
              u'op': u'insert',
              u'opid': 7,
              u'threadId': u'0x7fd98f626b40',
              u'waitingForLock': False},
             {u'active': False,
              u'desc': u'journal',
              u'lockStats': {u'timeAcquiringMicros': {u'R': 1559795996L,
                                                      u'W': 146453621L},
                             u'timeLockedMicros': {u'R': 8067580L,
                                                   u'W': 252379762L}},
              u'ns': u'',
              u'numYields': 0,
              u'op': u'none',
              u'opid': 2,
              u'query': {},
              u'threadId': u'0x7fd98a9a0700',
              u'waitingForLock': False},
             {u'active': False,
              u'desc': u'snapshotthread',
              u'lockStats': {u'timeAcquiringMicros': {},
                             u'timeLockedMicros': {}},
              u'ns': u'',
              u'numYields': 0,
              u'op': u'none',
              u'opid': 3,
              u'query': {},
              u'threadId': u'0x7fd989f9f700',
              u'waitingForLock': False},
             {u'active': False,
              u'desc': u'clientcursormon',
              u'lockStats': {u'timeAcquiringMicros': {},
                             u'timeLockedMicros': {}},
              u'ns': u'',
              u'numYields': 0,
              u'op': u'none',
              u'opid': 4,
              u'query': {},
              u'threadId': u'0x7fd98959e700',
              u'waitingForLock': False},
             {u'active': False,
              u'client': u'0.0.0.0:0',
              u'desc': u'TTLMonitor',
              u'lockStats': {u'timeAcquiringMicros': {u'r': 1L, u'w': 0L},
                             u'timeLockedMicros': {u'r': 94L, u'w': 0L}},
              u'ns': u'local.system.indexes',
              u'numYields': 0,
              u'op': u'query',
              u'opid': 3628920,
              u'query': {u'expireAfterSeconds': {u'$exists': True}},
              u'threadId': u'0x7fd98819c700',
              u'waitingForLock': False},
             {u'active': False,
              u'desc': u'websvr',
              u'lockStats': {u'timeAcquiringMicros': {},
                             u'timeLockedMicros': {}},
              u'ns': u'',
              u'numYields': 0,
              u'op': u'none',
              u'opid': 8,
              u'query': {},
              u'threadId': u'0x7fd98179b700',
              u'waitingForLock': False},
             {u'active': False,
              u'client': u'127.0.0.1:39610',
              u'connectionId': 2366,
              u'desc': u'conn2366',
              u'lockStats': {u'timeAcquiringMicros': {u'r': 4L, u'w': 0L},
                             u'timeLockedMicros': {u'r': 17523L, u'w': 0L}},
              u'ns': u'fishtest_new.system.indexes',
              u'numYields': 0,
              u'op': u'query',
              u'opid': 3629086,
              u'query': {u'ns': u'fishtest_new.runs'},
              u'threadId': u'0x7fd980d9a700',
              u'waitingForLock': False}]}
```

tomtor commented:
@xoto10 @ppigazzini Interesting, this shows that there are a number of queries concurrently in flight from different port numbers. Mainly user queries to verify the password (I am listed twice :) and a 'runs' query.

ppigazzini commented:
@Stefano80 I already wrote by hand this script in the `x` folder, it's working fine now, should we merge the PR?

Stefano80 commented:
It looks ok for me, let us merge it if it is useful.

tomtor commented:
Yes let's merge this

ppigazzini commented:
official server updated.
thank you @xoto10 !
